### PR TITLE
Introduce ExperimentalBlockEditorProvider

### PR DESCRIPTION
**Pull request by adamziel**
## Description

This PR introduces `ExperimentalBlockEditorProvider` that prevents mixing experimental editor settings and the public `BlockEditorProvider` component. The actual filtering is handled in a private `__experimentalUpdateSettings` selector in the block-editor store.

Experimental settings may still be set via a PHP plugin – limiting this vector would require a separate PR.

## Rationale

WordPress extenders cannot update the experimental block settings on their own. The `updateSettings()` actions of the `@wordpress/block-editor` store will filter out all the settings that are **not** a part of the public API. The only way to actually store them is via private action. `__experimentalUpdateSettings()`.

To privatize a block editor setting, add it to the `privateSettings` list in 
[/packages/block-editor/src/store/actions.js](/packages/block-editor/src/store/actions.js):

```js
const privateSettings = [
	'__unstableInserterMediaCategories',
	// List a block editor setting here to make it private
];
```

## Mobile apps

This PR updates a few `.native.js` files. Mobile apps choose to import the `.native.js` files over their `.js` counterparts whenever possible. At the same time, this PR introduced the expectation of having two extra named exports: `ExperimentalEditorProvider` and `ExperimentalBlockEditorProvider`. I updated the native files to export the regular Provider component under the experimental name. This works because the settings filtering is restricted to the web – it doesn't make sense in context of mobile apps anyway.

## Testing instructions

* This PR comes with unit tests – confirm they pass and no other CI checks are red
* Confirm the code changes make sense

cc @ntsekouras @youknowriad

### CODE REVIEW

File: packages/editor/src/index.js [hunk 1]
```diff
@@ -11,3 +11,14 @@ export * from './utils';
...
  */
 export { transformStyles } from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import { ExperimentalEditorProvider } from './components/provider';
+import { lock } from './experiments';
+
+export const experiments = {};
```
Comment (adamziel) [inline packages/editor/src/index.js]: This export is in here and not in experiments.js to avoid a circular dependency between `experiments.js` and `./components/provider/index.ts`

File: packages/block-editor/src/components/provider/index.js [hunk 2]
```diff
@@ -14,21 +14,81 @@ import { BlockRefsProvider } from './block-refs-provider';
...
+	'__experimentalUserCanCreatePages',
+	'__experimentalPreferPatternsOnRoot',
+	'__unstableIsPreviewMode',
+];
+const BlockEditorProvider = ( props ) => {
+	const cleanSettings = useMemo( () => {
+		const clean = {};
+		for ( const key in props.settings ) {
+			if (
+				key.startsWith( '__' ) &&
```
Comment (noisysocks) [inline packages/block-editor/src/components/provider/index.js]: I think we decided to not use the `__experimental` prefix for experimental props on a `lock()`ed component, yeah? So this check isn't sufficient as it won't remove experimental props with no `__` prefix.

File: packages/editor/src/index.js (hunk 1 above)
Comment (noisysocks) [reply packages/editor/src/index.js]: I'd prefer this to be consistent across packages so maybe we should export `experiments` from `index.js` in all the other packages too?

Or split `experiments.js` into `experiments.js` and `lock-unlock.js`.

File: packages/block-editor/src/components/provider/index.js (hunk 2 above)
Comment (adamziel) [reply packages/block-editor/src/components/provider/index.js]: In this case it's not merely about the component props, but about the block editor settings. I don't think the same logic applies. What do you think? If we omit the `__experimental` prefix on those then this check will need to go check the incoming props against through an exhaustice allow/denylist of experimental properties.

Comment (noisysocks) [reply packages/block-editor/src/components/provider/index.js]: I'm not sure 😅 just trying to be consistent. It would be confusing to know which settings are experimental vs. which ones aren't if we do not have the `__experimental` prefix.

At any rate since they will be private we can change our minds about this later.

File: packages/customize-widgets/src/experiments.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+/**
+ * WordPress dependencies
+ */
+import { __dangerousOptInToUnstableAPIsOnlyForCoreModules } from '@wordpress/experiments';
+
+export const { lock, unlock } =
+	__dangerousOptInToUnstableAPIsOnlyForCoreModules(
+		'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.',
+		'@wordpress/edit-site'
```
Comment (ntsekouras) [inline packages/customize-widgets/src/experiments.js]:
```suggestion
		'@wordpress/customize-widgets'
```

File: packages/block-editor/src/components/provider/index.js (hunk 2 above)
Comment (adamziel) [reply packages/block-editor/src/components/provider/index.js]: The basic algorithm is `if( isExperimental && isAddedAfter6.2) remove();`.

The difference we're discussing is between:

```js
const isExperimental = key.startsWith('__');
```

And:

```js
const isExperimental = aListOfAllExperimentalProperties.includes( key );
```

The latter seems high-maintenance and error-prone. It's super easy to forget about adding a property to the list when not doing so doesn't trigger any errors.

What if that dynamics was reverted, though? I'm thinking about:

```js
const experimentalSettings = [ /* A list of all the experimental settings */ ]
const stableSettings = [ /* A list of all the stable settings */ ]
if(!experimentalSettings.includes(key) && !stableSettings.includes(key)) {
    throw new Error('Every editor setting must be listed either in experimentalSettings or stableSettings');
}
const isExperimental = experimentalSettings.includes( key );
```

I would feel much more comfortable with that – the decision is clear-cut and enforced. What do you think @noisysocks ?

File: packages/editor/src/index.js (hunk 1 above)
Comment (adamziel) [reply packages/editor/src/index.js]: Agreed, let's make it consistent across packages. That's out of scope here, though – let's do it in a follow up PR

File: packages/block-editor/src/components/provider/index.js (hunk 2 above)
Comment (adamziel) [reply packages/block-editor/src/components/provider/index.js]: Actually... can WordPress extenders provide their own custom editor settings? If yes, then we cannot list all accepted stable and experimental settings! :-(

Comment (adamziel) [reply packages/block-editor/src/components/provider/index.js]: There is a filter:

```php
add_filter( 'block_editor_settings_all', 'gutenberg_get_block_editor_settings', PHP_INT_MAX );
```

What's more, there is no schema to enforce a specific shape of the settings object. You can add any settings you want to. An allowlist won't work here, then. Too bad.

It's down to the `__` prefix and a risky `aListOfAllExperimentalProperties`. I'll move forward with the first one in this PR, and if that turns out to be a problem then we can always revisit (as you mentioned).

Comment (ntsekouras) [reply packages/block-editor/src/components/provider/index.js]: What about only having a list with the settings we consider experimental. I know it's a bit error-prone as someone might forget to add it here, but doesn't rely on any naming convention and extenders could actually use whatever they want.

I'll think about it a bit more though :)

Comment (ntsekouras) [reply packages/block-editor/src/components/provider/index.js]: It would be good to add this to the docs, that we need to also add it this array.

**Comment by adamziel** [comment]
I added unit tests. There's still a problem with the native tests.

**Comment by adamziel** [comment]
This PR is blocked by https://github.com/WordPress/gutenberg/pull/47421

File: packages/block-editor/src/components/provider/test/use-block-sync.js [hunk 4]
```diff
@@ -14,7 +14,17 @@ import { render } from '@testing-library/react';
 import useBlockSync from '../use-block-sync';
 import withRegistryProvider from '../with-registry-provider';
 import * as blockEditorActions from '../../../store/actions';
+
 import { store as blockEditorStore } from '../../../store';
+jest.mock( '../../../store/actions', () => {
```
Comment (ntsekouras) [inline packages/block-editor/src/components/provider/test/use-block-sync.js:19]: Why this change was needed?

Comment (adamziel) [reply packages/block-editor/src/components/provider/test/use-block-sync.js:19]: Because of a nuance of importing – we used to spy on the actions module entry by replacing the `module[actionName]` via `jest.spyOn`. That was fine, but now there's an additional destructuring happening in `store.js`:

```js
import * as allActions from './actions';
const { __experimentalUpdateSettings, ...actions } = allActions;
```

The `spyOn` call was executed after the destructuring and had no effect on the registered store. `jest.mock()` call is, I believe, hoisted and occurs before the destructuring above – which means it has an effect on the store.

File: packages/data/src/registry.js [hunk 5]
```diff
@@ -297,6 +297,19 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 		const store = createReduxStore( storeName, options ).instantiate(
 			registry
 		);
+		// If creating a sub store, we need to copy the private actions and selectors
```
Comment (ntsekouras) [inline packages/data/src/registry.js]: leftover

**Review by ntsekouras — APPROVED** [review]
This LGTM! We'll wait for mobile folks to test the `native` files' changes and we can 🚢 .

Thanks Adam!

**Review by jhnstn — APPROVED** [review]
Changes look good. I didn't notice any issues running the changes in the mobile demo app.
